content/general/concepts/programming-optimization/programming-optimization.md

```diff
@@ -0,0 +1,265 @@
+
+Abstract data types tend to be more efficient for these optimizations.
+
+However, changing algorithms is easier than data structures and memoization is also an important dynamic programming caching technique to optimize the code from redundant computations.
+
+Because of the importance of caching, there are often many levels of caching in a system because of possible problems from memory use and correctness issues from stale caches.
+
+Beyond asymptotic order of growth, the constant factors matter: An asymptotically slower algorithm may be faster or smaller because it is simpler than an asymptotically faster algorithm when they are both faced with a small input, which may be the reality for most cases. 
+
+For example, a linear search is asymptotically slower than a binary search, but a linear search is faster for small inputs and a binary search is faster for large inputs.
+
+WOW, let's take a breather! 
+
+Hopefully haven't lost you yet, it is definitely a topic with many nuances. Ok, let's get back to it!
+
+### Source Code
+Next crucial level to wrap our minds around and can have a major impact on performance is the actual source code language implemented. 
+
+Having said that, each level is equally crucial since they all work in unison. 
+
+OK, here we go.
+
+Beyond algorithm implementation, source code level choices can make a significant impact in *programming optimization performance*. 
+
+Let's us throw an example out there.  
+
+*Control flow statements like while loops are slower than for loops for an unconditional loop. On early C compilers, for loops had an unconditional jump, however, while loops have a conditional jump for true or not verification.*
+
+That’s why their performance is different from early compilers. 
+
+Such aforementioned programming optimizations can now be performed by compilers. Go,Go tech! 
+
+In the most general terms, source code optimization tends to substitute generic high-level programming structures with lower-level source code, thus enhancing performance and/or reducing resource consumption(e.g. cpu, memory, etc.).
+
+Although compilers can optimize source code at a lower level, they are often limited to general purpose optimization, thus making manual optimization still predominant, allowing talented and experienced programmers to be creative with their solutions.  
+
+Important source code optimizing objectives to focus on would be: 
+- Rewrite or remove redundant code without disrupting the meaning of the now performing application. 
+- Reduce execution speed. 
+- Reduce available resources, whenever possible.
+
+Also important key points to keep in mind before optimizing the source code: 
+- Time based optimization can execute faster output, but it can also lead to a bigger code base, which can create conflict with memory and consumption. Consideration on the requirements needed will play an important role on this one.
```

Comment:
```suggestion
- Time based optimization can execute faster output, but it can also lead to a bigger code base, which can create conflict with memory and consumption.
```

Reply:
You marked this resolved without making the edit.